src/main/java/bitflow4j/script/registry/RegistryConstructor.java

```diff
@@ -183,31 +183,33 @@ public boolean register(Registry registry) {
     }
 
     private <T> RegisteredStep<T> configureRegisteredStep(RegisteredStep<T> reg) {
-        String[] constructor0ParamNames = paranamer.lookupParameterNames(constructors.get(0), false);
-        Set<String> optionalParams = new HashSet<>();
-        Set<String> requiredParams = new HashSet<>(Arrays.asList(constructor0ParamNames));
-
-        constructors.forEach(constructor -> {
-            List<String> paramNames = Arrays.asList(paranamer.lookupParameterNames(constructor, false));
-
-            // build optional parameters
-            for (String paramName : paramNames) {
-                if (!requiredParams.contains(paramName)) {
-                    optionalParams.add(paramName);
+        if (!constructors.isEmpty()) {
+            String[] constructor0ParamNames = paranamer.lookupParameterNames(constructors.get(0), false);
+            Set<String> optionalParams = new HashSet<>();
+            Set<String> requiredParams = new HashSet<>(Arrays.asList(constructor0ParamNames));
+
+            constructors.forEach(constructor -> {
+                List<String> paramNames = Arrays.asList(paranamer.lookupParameterNames(constructor, false));
+
+                // build optional parameters
+                for (String paramName : paramNames) {
+                    if (!requiredParams.contains(paramName)) {
+                        optionalParams.add(paramName);
+                    }
                 }
-            }
 
-            // clean requiredParams by moving extraneous parameters from requiredParams to optionalParams
-            for (Iterator<String> i = requiredParams.iterator(); i.hasNext(); ) {
-                String element = i.next();
-                if (!paramNames.contains(element)) {
-                    optionalParams.add(element);
-                    i.remove();
+                // clean requiredParams by moving extraneous parameters from requiredParams to optionalParams
+                for (Iterator<String> i = requiredParams.iterator(); i.hasNext(); ) {
+                    String element = i.next();
+                    if (!paramNames.contains(element)) {
+                        optionalParams.add(element);
+                        i.remove();
+                    }
                 }
-            }
-        });
-        reg.optional(optionalParams.toArray(new String[0]));
-        reg.required(requiredParams.toArray(new String[0]));
+            });
+            reg.optional(optionalParams.toArray(new String[0]));
```

Comment:
does it has to be new String[optionalParams.size()]  ?

Reply:
You are right, I think your way would be more efficient, because the created String array would be used by the toArray method directly, instead of allocating a new array. This way, the empty String array only serves as a "marker" that tells toArray() what kind of array to instantiate. It does not have a big effect though, because this code runs only once in the beginning of the program.